crates/client/eth/src/l1_gas_price.rs

```diff
@@ -0,0 +1,235 @@
+use alloy::eips::BlockNumberOrTag;
+use alloy::providers::Provider;
+use anyhow::Context;
+use dc_mempool::L1DataProvider;
+use futures::lock::Mutex;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+
+use lazy_static;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+const DEFAULT_GAS_PRICE_POLL_MS: u64 = 10_000;
+
+lazy_static::lazy_static! {
+    static ref LAST_UPDATE_TIMESTAMP: Arc<Mutex<u128>> = Arc::new(Mutex::new(0));
+}
+
+// Function to update the last update timestamp
+pub async fn update_last_update_timestamp() {
+    let now = SystemTime::now().duration_since(UNIX_EPOCH).expect("Time went backwards").as_millis();
+    let mut timestamp = LAST_UPDATE_TIMESTAMP.lock().await;
+    *timestamp = now;
+}
+
+// Function to get the last update timestamp
+pub async fn get_last_update_timestamp() -> u128 {
+    *LAST_UPDATE_TIMESTAMP.lock().await
+}
+
+pub async fn gas_price_worker(
+    eth_client: &EthereumClient,
+    l1_data_provider: Arc<dyn L1DataProvider>,
+    infinite_loop: bool,
+) -> anyhow::Result<()> {
+    let poll_time = eth_client.gas_price_poll_ms.unwrap_or(DEFAULT_GAS_PRICE_POLL_MS);
+    update_last_update_timestamp().await;
+    loop {
+        match update_gas_price(eth_client, Arc::clone(&l1_data_provider)).await {
+            Ok(_) => log::trace!("Updated gas prices"),
+            Err(e) => log::error!("Failed to update gas prices: {:?}", e),
+        }
+
+        let last_update_timestamp = get_last_update_timestamp().await;
+        let current_timestamp = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("Failed to get current timestamp")
+            .as_millis();
+
+        if current_timestamp - last_update_timestamp > 10 * poll_time as u128 {
+            panic!(
+                "Gas prices have not been updated for {} ms. Last update was at {}",
+                current_timestamp - last_update_timestamp,
+                last_update_timestamp
+            );
+        }
+
+        if !infinite_loop {
+            return Ok(());
+        }
+
+        sleep(Duration::from_millis(poll_time)).await;
+    }
+}
+
+async fn update_gas_price(
+    eth_client: &EthereumClient,
+    l1_data_provider: Arc<dyn L1DataProvider>,
+) -> anyhow::Result<()> {
+    let fee_history = eth_client.provider.get_fee_history(300, BlockNumberOrTag::Latest, &[]).await?;
+
+    // The RPC responds with 301 elements for some reason. It's also just safer to manually
+    // take the last 300. We choose 300 to get average gas caprice for last one hour (300 * 12 sec block
+    // time).
+    let (_, blob_fee_history_one_hour) =
+        fee_history.base_fee_per_blob_gas.split_at(fee_history.base_fee_per_blob_gas.len().max(300) - 300);
+
+    let avg_blob_base_fee = blob_fee_history_one_hour.iter().sum::<u128>() / blob_fee_history_one_hour.len() as u128;
+
+    let eth_gas_price = fee_history.base_fee_per_blob_gas.last().context("Setting l1 last confirmed block number")?;
```

Comment:
```suggestion
    let eth_gas_price = fee_history.base_fee_gas.last().context("Getting eth gas price")?;
```

Reply:
anyway we can ensure this in test cases? I think anvil allows you to set a custom gas price when you start it

Reply:
yeah that's true, I think we can achieve this using mocks right?

Reply:
resolved